src/librustc_lint/types.rs

```diff
@@ -957,8 +975,7 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         // it is only OK to use this function because extern fns cannot have
         // any generic types right now:
         let ty = self.cx.tcx.normalize_erasing_regions(ParamEnv::reveal_all(), ty);
-
-        match self.check_type_for_ffi(&mut FxHashSet::default(), ty) {
+        match self.check_type_for_ffi(ty) {
```

Comment:
Wait, I'm confused, I thought there were two callsites before (one for returns and one for)? I must've misread the previous version of this, I didn't realize there was a single callsite.

In which case, the array check could just be here, kind of like `check_for_opaque_ty`, but after `normalize_erasing_regions`.

Reply:
Oh and you have to find a way to make the call from `check_foreign_static` not trigger the array check (I forgot we also check `static`s, sorry).

Reply:
Oh, I haven't though of that (and didn't see any tests, i'll add some of those too)

Reply:
Tried thinking of a better way to do this, but the least destructive way is to pass a bool `is_static` 
otherwise we need to start passing `hir::ForeignItem` down the chain to check if static